File: DGV7.0-NON-SAAS/PayHub-Guest-iOS/PayHubGuest/Models/GuestHistoryStore.swift
import Foundation

/// Guest Mode has no login/session, so there is nothing on the server to page through — history
/// is cached on-device only, as JSON in UserDefaults, keyed by transaction reference.
enum GuestHistoryStore {
    private static let key = "guest_history_receipts"
    private static let maxEntries = 50

    static func load() -> [GuestReceipt] {
        guard let data = UserDefaults.standard.data(forKey: key) else { return [] }
        guard let receipts = try? JSONDecoder().decode([GuestReceipt].self, from: data) else { return [] }
        return receipts.sorted { $0.date > $1.date }
    }

    @discardableResult
    static func save(_ receipt: GuestReceipt) -> [GuestReceipt] {
        var deduped = load().filter { $0.reference != receipt.reference }
        deduped.insert(receipt, at: 0)
        let updated = Array(deduped.sorted { $0.date > $1.date }.prefix(maxEntries))
        if let data = try? JSONEncoder().encode(updated) {
            UserDefaults.standard.set(data, forKey: key)
        }
        return updated
    }
}

File: DGV7.0-NON-SAAS/PayHub-Guest-Android/app/src/main/java/com/payhub/guest/data/GuestHistoryStore.kt
package com.payhub.guest.data

import android.content.Context
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.payhub.guest.data.model.GuestReceipt

/**
 * Guest Mode has no login/session, so there is nothing on the server to page through — history
 * is cached on-device only, as a JSON list in SharedPreferences, keyed by transaction reference.
 */
object GuestHistoryStore {
    private const val PREFS_NAME = "guest_history"
    private const val KEY_RECEIPTS = "receipts"
    private const val MAX_ENTRIES = 50

    private val gson = Gson()
    private val listType = object : TypeToken<List<GuestReceipt>>() {}.type

    fun load(context: Context): List<GuestReceipt> {
        val json = prefs(context).getString(KEY_RECEIPTS, null) ?: return emptyList()
        return try {
            gson.fromJson<List<GuestReceipt>>(json, listType)?.sortedByDescending { it.dateMillis } ?: emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun save(context: Context, receipt: GuestReceipt): List<GuestReceipt> {
        val deduped = load(context).filterNot { it.reference == receipt.reference }
        val updated = (listOf(receipt) + deduped).sortedByDescending { it.dateMillis }.take(MAX_ENTRIES)
        prefs(context).edit().putString(KEY_RECEIPTS, gson.toJson(updated)).apply()
        return updated
    }

    private fun prefs(context: Context) = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
